providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -94,10 +93,23 @@ static void *mlkem_new(void *provctx)
     debug_print("MLKEMKM new key req\n");
     if (!ossl_prov_is_running())
         return 0;
+
     key = OPENSSL_zalloc(sizeof(MLKEM768_KEY));
-    if (key == NULL)
-        return 0;
-    key->keytype = MLKEM_KEY_TYPE_768; /* TODO(ML-KEM) any type */
+    if (key != NULL) {
+        key->keytype = MLKEM_KEY_TYPE_768; /* TODO(ML-KEM) any type */
+        key->provctx = provctx;
+        /*
+         * ideally, this is a one-time allocation and ctx that should be within the
+         * provider context: OK to move it there?? It would be the first algorithm-
+         * specific context stored: Feels weird (TODO(ML-KEM)).
+         */
```

Comment:
This is debatable. In theory the preferred implementation of these SHA3/SHAKE hashes in the context can change for example by setting the default propq. Caching the fetched implementations within the provider context would made these fixed forever. Also the cache access would require mutex for initialization unless we would cache it immediately on provider load which is also slightly problematic. IMO not worth it.

Reply:
For now OK to leave the comment so we can come back to that later/before final merge?

Reply:
IMO the comment is not worth leaving, but is this is my personal opinion.

Reply:
> Caching the fetched implementations within the provider context would made these fixed forever. 

"forever" is a rather relative term in this context, no? Basically, "forever" as long as default provider is configured/loaded. Given these algorithms are very much a core element of proper operation of ML-KEM, I'd see it as beneficial that they cannot be changed at runtime/once default provider initialization is complete.

> unless we would cache it immediately on provider load which is also slightly problematic

That would've been my preference -- so what problem do you see with that @t8m?

Reply:
Let's imagine a user wants to replace the implementations with some hw-accelerator based ones in a different provider - this provider might not be present in the libctx during the default provider load.

Reply:
Well, then that user could (even should?) change the provider initialization order as per his setup, no? I'm personally not too happy about giving up performance in the common case/for nearly everyone for an corner case that someone creates and can work around by a different setup. But then again, performance is so good right now (at least compared to oqsprovider) that we can worry about that later.

Reply:
OK -- but it seems some optimization would be prudent looking at the most recent like-for-like comparisons (admittedly not quite fair running AVX2 optimized code against pure C code).